cpp/contracts/adapter.cpp

```diff
@@ -457,50 +454,48 @@ void pam::check_authorization(name adapter, const operation& operation, const me
 
    bytes topic_zero = extract_32bytes(event_payload, offset);
    check(topic_zero == exp_topic_zero && !is_all_zeros(topic_zero), "unexpected Topic Zero");
-   offset += 32 * 3; // skip other topics
+   offset += 32;
 
-   // check nonce
-   bytes event_data(event_payload.begin() + offset, event_payload.end());
-   bytes nonce = extract_32bytes(event_data, offset);
+   bytes nonce = extract_32bytes(event_payload, offset);
    uint64_t nonce_int = bytes32_to_uint64(nonce);
    check(operation.nonce == nonce_int, "nonce do not match");
-   offset += 32;
+   offset += 32 * 3; // skip other topics
```

Comment:
I see two errors here:
 1. You are expecting a nonce always within the topics which is not true in reality (what if the event is coming from another eosio chain?)
 2. After the nonce (which is at topic1 position) you are skipping 3 topics again while you should skip 2 of them

I suppose 2. is still working because there is the operation.nonce just after the topics, but still the comment (skip topics) is misleading.

Anyway, if you don't have anything against it, I think you should revert this change and pick the operation.nonce as done before

Reply:
Using the first metadata sample which is generated starting from a EVM swap 
```
{
  "address": "0xa899118f4bccb62f8c6a37887a4f450d8a4e92e0",
  "topics": [
    "0x66756e6473206172652073616675207361667520736166752073616675202e2e",
    "0x0000000000000000000000000000000000000000000000000000000000000001"
  ],
  "data": "0x000000000000000000000000e58cbe144dd5556c84874dec1b3f2d0d6ac45f1baca376f206b8fc25a6ed44dbdc66547c36c6c33e3a119ffbeaef943642f0e906000000000000000000000000000000000000000000000000517d3388612e7c00000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266000000000000000000000000000000000000000000000000000000000000000c656f73726563697069656e74",
  "blockHash": "0x0e3595e678db71061a7fcba915f9ffc91f84cc0f5dee8e3cdfee83a68793d09b",
  "blockNumber": "0x14188ec",
  "blockTimestamp": "0x67210819",
  "transactionHash": "0xbe5c527a4c59a275b7493fab9984e26c445256dfbd328fcb928c6925baaf326c",
  "transactionIndex": "0x0",
  "logIndex": "0x7",
  "removed": false
}
```
We have this preimage:
```
// context
0101                                                             // offset +2
0000000000000000000000000000000000000000000000000000000000000001 // extract 32 bytes -> origin chain id 
// tx infos
0e3595e678db71061a7fcba915f9ffc91f84cc0f5dee8e3cdfee83a68793d09b // block id
be5c527a4c59a275b7493fab9984e26c445256dfbd328fcb928c6925baaf326c // tx id
/*
 * @event_payload
 * offset = 0; 
 * bytes event_payload(metadata.preimage.begin() + 98, metadata.preimage.end());
 */
000000000000000000000000a899118f4bccb62f8c6a37887a4f450d8a4e92e0 // emitter
66756e6473206172652073616675207361667520736166752073616675202e2e // topic_zero (topic 0)
0000000000000000000000000000000000000000000000000000000000000000 // nonce (topic 1)
0000000000000000000000000000000000000000000000000000000000000000 // (topic 2)
0000000000000000000000000000000000000000000000000000000000000000 // (topic 3)
000000000000000000000000e58cbe144dd5556c84874dec1b3f2d0d6ac45f1b // token
aca376f206b8fc25a6ed44dbdc66547c36c6c33e3a119ffbeaef943642f0e906 // dest_chain_id
000000000000000000000000000000000000000000000000517d3388612e7c00 // amount
000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266 // sender
000000000000000000000000000000000000000000000000000000000000000c // recipient_len
656f73726563697069656e74                                         // recipient
```
There are actually 4 topics, they are introduced at https://github.com/pnetwork-association/pnetwork-v4/blob/master/javascript/event-attestator/src/ProofcastEventAttestator.js#L65

We can avoid this, but I kept the eventAttestator.js as reference.

Reply:
I think you are misaligned then, sorry I didn't update see https://github.com/pnetwork-association/pnetwork-v4/pull/50

Reply:
ok, clear now.
What about the point 2? The preimage introduces 4 bytes anyway.

Reply:
> I would remove all topics in the preimage and keep only the data field, given the previous point.

I don't think it's a good idea, maybe it would be good for pNetwork contracts but not in general (topic zero identify the name of the event in general, keep in mind that most contracts use `emit Event` instead, not `log2`)